velocity/src/main/java/com/github/games647/fastlogin/velocity/listener/ConnectListener.java

```diff
@@ -0,0 +1,130 @@
+            return;
+        }
+
+        String username = preLoginEvent.getUsername();
+        plugin.getLog().info("Incoming login request for {} from {}", username, connection.getRemoteAddress());
+
+        Runnable asyncPremiumCheck = new AsyncPremiumCheck(plugin, connection, username, continuation, preLoginEvent);
+        plugin.getScheduler().runAsync(asyncPremiumCheck);
+    }
+
+    @Subscribe
+    public void onGameprofileRequest(GameProfileRequestEvent event) {
+        if (event.isOnlineMode()) {
+            LoginSession session = plugin.getSession().get(event.getConnection().getRemoteAddress());
+
+            UUID verifiedUUID = event.getGameProfile().getId();
+            String verifiedUsername = event.getUsername();
+            session.setUuid(verifiedUUID);
+            session.setVerifiedUsername(verifiedUsername);
+
+            StoredProfile playerProfile = session.getProfile();
+            playerProfile.setId(verifiedUUID);
+            if (!plugin.getCore().getConfig().get("premiumUuid", true)) {
+                UUID offlineUUID = UUIDAdapter.generateOfflineId(playerProfile.getName());
+                event.setGameProfile(event.getGameProfile().withId(offlineUUID));
+                plugin.getLog().info("Overridden UUID from {} to {} (based of {}) on {}",
+                        verifiedUUID, offlineUUID, verifiedUsername, event.getConnection());
+            }
+
+            if (!plugin.getCore().getConfig().get("forwardSkin", true)) {
+                //FIXME: Do I need to remove *all* properties or only the skin related ones?
+                event.setGameProfile(event.getGameProfile().withProperties(new ArrayList<>()));
+            }
+        }
+    }
+
+    @Subscribe
+    public void onServerConnected(ServerConnectedEvent serverConnectedEvent) {
+        Player player = serverConnectedEvent.getPlayer();
+        RegisteredServer server = serverConnectedEvent.getServer();
+
+        VelocityLoginSession session = plugin.getSession().get(player.getRemoteAddress());
+        if (session == null) {
+            return;
+        }
+
+        // delay sending force command, because Paper will process the login event asynchronously
+        // In this case it means that the force command (plugin message) is already received and processed while
+        // player is still in the login phase and reported to be offline.
+        Runnable loginTask = new ForceLoginTask(plugin.getCore(), player, server, session);
+        plugin.getProxy().getScheduler()
+                .buildTask(plugin, () -> plugin.getScheduler().runAsync(loginTask))
+                .delay(1L, TimeUnit.SECONDS) // Delay at least one second, otherwise the login command can be missed
+                .schedule();
+    }
+
+    @Subscribe
+    public void onDisconnect(DisconnectEvent disconnectEvent) {
+        Player player = disconnectEvent.getPlayer();
+        assert plugin.getSession().remove(player.getRemoteAddress()) != null;
```

Comment:
Correct me if I am wrong, but asserts doesn't work in without special runtime parameters. This is better solved in other programming languages ;)

Reply:
That was a left over from when I was testing for a suitable object for the player session. I'll be removing it.